flipt-engine-ffi/src/http.rs

```diff
@@ -148,207 +213,311 @@ impl Parser for HTTPParser {
+                if let Some(location) = response.headers().get(reqwest::header::LOCATION) {
+                    self.url = location.to_str().unwrap().to_string();
+                    self.mode = FetchMode::Streaming;
+                }
+                Ok(Some(response))
+            }
+
             reqwest::StatusCode::OK => {
                 // check if we have a new etag
                 if let Some(etag) = response.headers().get(reqwest::header::ETAG) {
                     self.etag = Some(etag.to_str().unwrap().to_string());
                 }
 
-                let response_text = match response.text() {
-                    Ok(t) => t,
-                    Err(e) => {
-                        return Err(Error::Server(format!("failed to get response body: {}", e)))
-                    }
-                };
-
-                match serde_json::from_str(&response_text) {
-                    Ok(doc) => Ok(Some(doc)),
-                    Err(e) => Err(Error::InvalidJSON(e.to_string())),
-                }
+                Ok(Some(response))
             }
+
             _ => Err(Error::Server(format!(
                 "unexpected http response: {} {}",
                 response.status(),
-                response.text().unwrap_or("".to_string())
+                response.text().await.unwrap_or("".to_string())
             ))),
         }
     }
+
+    async fn fetch_stream(&mut self) -> Result<Option<Response>, Error> {
+        self.http_client
+            .get(self.url.clone())
+            .headers(self.build_headers())
+            .send()
+            .await
+            .map_err(|e| {
+                self.stop();
+                Error::Server(format!("failed to make request: {}", e))
+            })?
+            .error_for_status()
+            .map_err(|e| {
+                self.stop();
+                Error::Server(format!("response: {}", e))
+            })
+            .map(Some)
+    }
+
+    async fn handle_polling(
+        &mut self,
+        sender: &mpsc::Sender<Result<source::Document, Error>>,
+    ) -> Result<(), Error> {
+        match self.fetch().await {
+            Ok(Some(resp)) => {
+                let text = resp
+                    .text()
+                    .await
+                    .map_err(|e| Error::Server(format!("failed to read response body: {}", e)))?;
+                let doc: source::Document = serde_json::from_str(&text).map_err(|e| {
+                    Error::InvalidJSON(format!("failed to parse response body: {}", e))
+                })?;
+                sender
+                    .send(Ok(doc))
+                    .await
+                    .map_err(|_| Error::Server("Failed to send result".into()))?;
+            }
+            Ok(None) => {}
+            Err(e) => sender
+                .send(Err(e))
+                .await
+                .map_err(|_| Error::Server("Failed to send error".into()))?,
+        }
+        Ok(())
```

Comment:
Do we want this to always evaluate to `OK()` ?
Seems like this might shadow any error in the previous expresson which the caller does match on in a `let` expression to print out.

Reply:
if there's any errors they get handled in the `map_err` calls and the `?` at the end means they get returned early to the caller:

```
.map_err(|_| Error::Server("Failed to send result".into()))?;
```